File: app/src/main/java/closer/vlllage/com/closer/ListActivity.java
package closer.vlllage.com.closer;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Animation;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.TranslateAnimation;

import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.pool.PoolActivity;

public abstract class ListActivity extends PoolActivity {

    protected View background;
    protected RecyclerView recyclerView;

    private TranslateAnimation finishAnimator;
    private Runnable finishCallback;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        overridePendingTransition(0, 0);
        setContentView(R.layout.activity_share);

        background = findViewById(R.id.background);
        recyclerView = findViewById(R.id.shareRecyclerView);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        background.setOnTouchListener((view, motionEvent) -> {
            background.setOnTouchListener(null);
            finish();
            return true;
        });

        ViewTreeObserver viewTreeObserver = background.getViewTreeObserver();
        viewTreeObserver.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                background.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                reveal();
            }

            private void reveal() {
                float yPercent = ($(ResourcesHandler.class).getResources().getDimensionPixelSize(R.dimen.feedPeekHeight) * 2) / (float) recyclerView.getMeasuredHeight();
                TranslateAnimation animation = new TranslateAnimation(
                        Animation.RELATIVE_TO_PARENT, 0, Animation.RELATIVE_TO_PARENT, 0, Animation.RELATIVE_TO_PARENT, yPercent, Animation.RELATIVE_TO_PARENT, 0
                );
                animation.setDuration(225);
                animation.setInterpolator(new DecelerateInterpolator());
                recyclerView.startAnimation(animation);
            }
        });
    }

    protected boolean isDone() {
        return finishAnimator != null;
    }

    @Override
    public void finish() {
        if (finishAnimator != null) {
            return;
        }

        finishAnimator = new TranslateAnimation(
                Animation.RELATIVE_TO_PARENT, 0, Animation.RELATIVE_TO_PARENT, 0, Animation.RELATIVE_TO_PARENT, 0, Animation.RELATIVE_TO_PARENT, 1
        );;
        finishAnimator.setDuration(195);
        finishAnimator.setInterpolator(new AccelerateDecelerateInterpolator());
        finishAnimator.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                ListActivity.super.finish();
                getWindow().getDecorView().setVisibility(View.GONE);
                overridePendingTransition(0, 0);
                finishAnimator = null;

                if (finishCallback != null) {
                    finishCallback.run();
                    finishCallback = null;
                }
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        recyclerView.startAnimation(finishAnimator);
    }

    public void finish(Runnable callback) {
        finishCallback = callback;
        finish();
    }
}

File: app/src/main/java/closer/vlllage/com/closer/ListActivity.kt
package closer.vlllage.com.closer

import android.animation.ArgbEvaluator
import android.animation.ObjectAnimator
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import android.view.ViewTreeObserver
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.Animation
import android.view.animation.DecelerateInterpolator
import android.view.animation.TranslateAnimation
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.extensions.visible
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler
import closer.vlllage.com.closer.pool.PoolActivity
import kotlinx.android.synthetic.main.activity_share.*

abstract class ListActivity : PoolActivity() {

    protected lateinit var recyclerView: RecyclerView

    private var finishAnimator: TranslateAnimation? = null
    private var finishCallback: Runnable? = null

    protected var closeCallback: (() -> Boolean)? = null

    protected val isDone: Boolean
        get() = finishAnimator != null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        overridePendingTransition(0, 0)
        setContentView(R.layout.activity_share)

        background.clipToOutline = true
        recyclerView = shareRecyclerView

        recyclerView.layoutManager = LinearLayoutManager(this)

        background.setOnTouchListener { view, motionEvent ->
            if (motionEvent.action == MotionEvent.ACTION_DOWN) {
                if (closeCallback?.invoke() != false) {
                    background.setOnTouchListener(null)
                    finish()
                }
                true
            } else false
        }

        val viewTreeObserver = background.viewTreeObserver
        viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                background.viewTreeObserver.removeOnGlobalLayoutListener(this)
                reveal()
            }

            private fun reveal() {
                val colorAnim = ObjectAnimator.ofInt(background, "backgroundColor", getColor(R.color.black_transparent), getColor(R.color.black_10))
                colorAnim.duration = 225
                colorAnim.setEvaluator(ArgbEvaluator())
                colorAnim.start()

                val yPercent = on<ResourcesHandler>().resources.getDimensionPixelSize(R.dimen.feedPeekHeight) * 2 / recyclerView.measuredHeight.toFloat()
                val animation = TranslateAnimation(
                        Animation.RELATIVE_TO_PARENT, 0f, Animation.RELATIVE_TO_PARENT, 0f, Animation.RELATIVE_TO_PARENT, yPercent, Animation.RELATIVE_TO_PARENT, 0f
                )
                animation.duration = 225
                animation.interpolator = DecelerateInterpolator()
                recyclerView.startAnimation(animation)
            }
        })
    }

    override fun finish() {
        if (finishAnimator != null) {
            return
        }

        val colorAnim = ObjectAnimator.ofInt(background, "backgroundColor", getColor(R.color.black_10), getColor(R.color.black_transparent))
        colorAnim.duration = 195
        colorAnim.setEvaluator(ArgbEvaluator())
        colorAnim.start()

        finishAnimator = TranslateAnimation(
                Animation.RELATIVE_TO_PARENT, 0f, Animation.RELATIVE_TO_PARENT, 0f, Animation.RELATIVE_TO_PARENT, 0f, Animation.RELATIVE_TO_PARENT, 1f
        )
        finishAnimator!!.duration = 195
        finishAnimator!!.interpolator = AccelerateDecelerateInterpolator()
        finishAnimator!!.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {

            }

            override fun onAnimationEnd(animation: Animation) {
                super@ListActivity.finish()
                window.decorView.visible = false
                overridePendingTransition(0, 0)
                finishAnimator = null

                finishCallback?.run()
            }

            override fun onAnimationRepeat(animation: Animation) {

            }
        })
        recyclerView.startAnimation(finishAnimator)
    }

    fun finish(callback: Runnable) {
        finishCallback = callback
        finish()
    }
}
